### PR TITLE
experiment fallback to tcp when udp unstable

### PR DESCRIPTION
**Pull request by cnderrauber**

### CODE REVIEW

File: pkg/rtc/transportmanager.go [hunk 1]
```diff
@@ -609,3 +623,32 @@ func (t *TransportManager) ProcessPendingPublisherDataChannels() {
 		}
 	}
 }
+
+func (t *TransportManager) OnReceiverReport(dt *sfu.DownTrack, report *rtcp.ReceiverReport) {
+	t.mediaLossProxy.HandleMaxLossFeedback(dt, report)
+}
+
+func (t *TransportManager) onMediaLossUpdate(loss uint8) {
+	t.udpLossUnstableCount <<= 1
```
Comment (boks1971) [inline pkg/rtc/transportmanager.go:632]: This needs to be locked? A lot of tracks could be flowing through this I guess.

File: pkg/rtc/transportmanager.go [hunk 2]
```diff
@@ -20,6 +21,12 @@ import (
 
 const (
 	failureCountThreshold = 2
+
+	// when RR report loss frac over this threshold, we consider it is a unstable event
```
Comment (boks1971) [inline pkg/rtc/transportmanager.go:25]: maybe modify comment here to say that this percentage (25%). There is conversion below to RR fraction lost scale where 25% will become 64 or so. As RTCP calls it loss fraction, suggesting comment here as percentage.

File: pkg/rtc/transportmanager.go [hunk 3]
```diff
@@ -609,3 +623,32 @@ func (t *TransportManager) ProcessPendingPublisherDataChannels() {
...
+			if t.udpRTT > 0 && t.tcpRTT < uint32(float32(t.udpRTT)*1.3) && t.tcpRTT < tcpGoodRTT {
+				t.udpLossUnstableCount = 0
+				t.params.Logger.Debugw("udp connection unstable, switch to tcp")
+				t.handleConnectionFailed(true)
+				if t.onAnyTransportFailed != nil {
+					t.onAnyTransportFailed()
+				}
+			}
+		}
+	}
```
Comment (boks1971) [inline pkg/rtc/transportmanager.go:645]: Should this function actually check that the transport is UDP? Maybe setup the `MediaLossProxy` callback only after the `SUBSCRIBER` ICE connection is established.

File: pkg/rtc/transportmanager.go [hunk 4]
```diff
@@ -609,3 +623,32 @@ func (t *TransportManager) ProcessPendingPublisherDataChannels() {
...
+func (t *TransportManager) OnReceiverReport(dt *sfu.DownTrack, report *rtcp.ReceiverReport) {
+	t.mediaLossProxy.HandleMaxLossFeedback(dt, report)
+}
+
+func (t *TransportManager) onMediaLossUpdate(loss uint8) {
+	t.udpLossUnstableCount <<= 1
+	if loss >= uint8(255*udpLossFracUnstable/100) {
+		t.udpLossUnstableCount |= 1
+		if bits.OnesCount32(t.udpLossUnstableCount) >= udpLossUnstableCountThreshold {
+			if t.udpRTT > 0 && t.tcpRTT < uint32(float32(t.udpRTT)*1.3) && t.tcpRTT < tcpGoodRTT {
```
Comment (boks1971) [inline pkg/rtc/transportmanager.go:636]: Should we actually check against `min` of UDP RTT? Or some smoothed version of it? This is checking against the latest RTT when the count crosses the threshold. Wondering if that is going to be erratic.

File: pkg/rtc/transportmanager.go [hunk 5]
```diff
@@ -20,6 +21,12 @@ import (
 
 const (
 	failureCountThreshold = 2
+
+	// when RR report loss frac over this threshold, we consider it is a unstable event
+	udpLossFracUnstable = 25
+	// if in last 32 times RR, the unstable report count over this threshold, the connection is unstable
+	udpLossUnstableCountThreshold = 20
+	tcpGoodRTT                    = 80
```
Comment (boks1971) [inline pkg/rtc/transportmanager.go:29]: Should we add a check for RTCP RR report with min number of packets. With DTX, there might be windows with small number of packets.

File: pkg/rtc/transportmanager.go (hunk 1 above)
Comment (cnderrauber) [reply pkg/rtc/transportmanager.go:632]: The MediaLossProxy has a minimum update interval (1 sec), so it is safe without lock

File: pkg/rtc/transportmanager.go (hunk 4 above)
Comment (cnderrauber) [reply pkg/rtc/transportmanager.go:636]: Added a smooth to RTT update.  rtt += (newRtt - oldRtt)/2

File: pkg/rtc/transportmanager.go (hunk 3 above)
Comment (boks1971) [reply pkg/rtc/transportmanager.go:645]: Okay. I was more thinking about the case of `preference = None`, but ICE chooses a different path. With TCP, there would be no loss. So, that should be okay. For TURN/TLS also, loss should be low even though the path between TURN server and SFU will be UDP.

File: pkg/rtc/transportmanager.go (hunk 1 above)
Comment (boks1971) [reply pkg/rtc/transportmanager.go:632]: Yeah. But, it is probably okay to lock as it is once a second. Just don't want code to behave unexpectedly if something else changes. So, if it is not critical path, have been advocating safety :-)

**Review by boks1971 — APPROVED** [review]

File: pkg/rtc/transportmanager.go (hunk 1 above)
Comment (cnderrauber) [reply pkg/rtc/transportmanager.go:633]: Done

File: pkg/rtc/transportmanager.go (hunk 5 above)
Comment (cnderrauber) [reply pkg/rtc/transportmanager.go:29]: Don't want to count packet numbers in transport manager, might need do it in other place(downtrack or medialossproxy), will draft another PR to refine that.

**Pull request merged**